File: Mathematics/Matrix.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Runtime.CompilerServices;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;

namespace Mathematics;

public class Matrix : IMatrix
{
    private float[,] xs;

    public int M_Rows => xs.GetLength(0);

    public int N_Cols => xs.GetLength(1);

    public (int, int) Size => (M_Rows, N_Cols);

    public Matrix(int mRows, int nCols)
    {
        xs = new float[mRows, nCols];
    }

    public Matrix(float[,] xs)
    {
        this.xs = xs;
    }

    public Matrix(Matrix A)
    {
        xs = new float[A.M_Rows, A.N_Cols];

        for(int i = 0; i <  A.M_Rows; i++)
            for(int j = 0; j < A.N_Cols; j++)
                xs[i, j] = A.xs[i, j];
    }

    public float Item(int i, int j) => xs[i, j];

    public void SetItem(int i, int j, float value)
    {
        xs[i, j] = value;
    }

    public Vector Row(int i)
    {
        Vector vec = new Vector(M_Rows);
        for (int j = 0; j < M_Rows; j++)
            vec.SetItem(j, Item(j, i));

        return vec;
    }

    public float[,] ToArray()
    {
        return xs;
    }

    public Vector Column(int j)
    {
        Vector vec = new Vector(N_Cols);
        
        for (int i = 0; i < N_Cols; i++)
            vec.SetItem(i, Item(j, i));

        return vec;
    }
}

File: FSharpMathematics/Core/Matrix.fs
namespace FSharpMathematics.Core

open FSharp.Collections

type Matrix = class
  val private _xs : float[,]

  /// <summary>
  /// Initializes an m-by-n Matrix with all 0's.
  /// </summary>
  new (mRows : int, nCols : int) =
    { _xs = Array2D.zeroCreate mRows nCols }

  /// <summary>
  /// Initialises a Matrix from a Array2D
  /// </summary>
  /// <remarks>
  /// the new object has its own instance of _xs
  /// </remarks> 
  new (xs : float[,]) =
    { _xs = Array2D.copy xs }
  

  ///<summary>
  /// A copy constructor
  ///</summary 
  /// <remarks>
  /// Deep copy, no instances shared.
  /// </remarks>
  new (A : Matrix) = 
    {_xs = Array2D.copy A._xs}


  /// <summary>
  /// Get the matrix as a 2D array.
  /// </summary>
  /// 
  /// <returns>
  /// A 2D array representation of the Matrix.
  /// </returns>
  member this.ToArray() =
    this._xs

  member this.M_Rows =
    this._xs.GetLength(0)

  member this.N_Cols =
    this._xs.GetLength(1)

  member this.Size =
    (this.M_Rows, this.N_Cols)

  member this.Item
    with get (i : int, j : int) = this._xs.[i, j]
    and set (i : int, j : int) (value : float) = this._xs.[i, j] <- value

  /// <summary>
  /// Get a particular row vector.
  /// </summary>
  ///
  /// <param name="i">
  /// The index of the row to get.
  /// </param>
  ///
  /// <returns>
  /// A vector containing the data of the given row
  /// </returns>
  member this.Row (i : int) : Vector =
    Vector(seq { for i in 0..this.N_Cols-1 -> i } |> Seq.map (fun x -> this._xs[i, x]) |> Seq.toArray)

  /// <summary>
  /// Get a particular column vector.
  /// </summary>
  ///
  /// <param name="j">
  /// The index of the column to get.
  /// </param>
  ///
  /// <returns>
  /// A vector containing the data of the given column
  /// </returns>
  ///
  /// <remarks>
  /// The implementation is row-major, so this is expected to be a slow
  /// endeavour.
  /// </remarks>
  member this.Column (j : int) =
    Vector (seq {for i in 0..this.M_Rows-1 -> i } |> Seq.map (fun x -> this._xs[x, j]) |> Seq.toArray)

end